From 913366d26c0b044755fd4d387abc1eaca8a8c4f0 Mon Sep 17 00:00:00 2001
From: Alexander Hofbauer <alex@derhofbauer.at>
Date: Sat, 7 Feb 2015 19:22:56 +0100
Subject: [PATCH] Emit a signal for stream-changed

---
 gvc-mixer-control.c | 29 +++++++++++++++++++++++++++++
 gvc-mixer-control.h |  2 ++
 2 files changed, 31 insertions(+)

diff --git a/gvc-mixer-control.c b/gvc-mixer-control.c
index b270523..f708dda 100644
--- a/gvc-mixer-control.c
+++ b/gvc-mixer-control.c
@@ -104,6 +104,7 @@ enum {
         STATE_CHANGED,
         STREAM_ADDED,
         STREAM_REMOVED,
+        STREAM_CHANGED,
         CARD_ADDED,
         CARD_REMOVED,
         DEFAULT_SINK_CHANGED,
@@ -1528,6 +1529,11 @@ update_sink (GvcMixerControl    *control,
                 /* Always sink on a new stream to able to assign the right stream id
                  * to the appropriate outputs (multiple potential outputs per stream). */
                 sync_devices (control, stream);
+        } else {
+                g_signal_emit (G_OBJECT (control),
+                               signals[STREAM_CHANGED],
+                               0,
+                               gvc_mixer_stream_get_id (stream));
         }
 
         /*
@@ -1648,6 +1654,11 @@ update_source (GvcMixerControl      *control,
                                      g_object_ref (stream));
                 add_stream (control, stream);
                 sync_devices (control, stream);
+        } else {
+                g_signal_emit (G_OBJECT (control),
+                               signals[STREAM_CHANGED],
+                               0,
+                               gvc_mixer_stream_get_id (stream));
         }
 
         if (control->priv->profile_swapping_device_id != GVC_MIXER_UI_DEVICE_INVALID) {
@@ -1753,6 +1764,11 @@ update_sink_input (GvcMixerControl          *control,
                                      GUINT_TO_POINTER (info->index),
                                      g_object_ref (stream));
                 add_stream (control, stream);
+        } else {
+                g_signal_emit (G_OBJECT (control),
+                               signals[STREAM_CHANGED],
+                               0,
+                               gvc_mixer_stream_get_id (stream));
         }
 }
 
@@ -1799,6 +1815,11 @@ update_source_output (GvcMixerControl             *control,
                                      GUINT_TO_POINTER (info->index),
                                      g_object_ref (stream));
                 add_stream (control, stream);
+        } else {
+                g_signal_emit (G_OBJECT (control),
+                               signals[STREAM_CHANGED],
+                               0,
+                               gvc_mixer_stream_get_id (stream));
         }
 }
 
@@ -3208,6 +3229,14 @@ gvc_mixer_control_class_init (GvcMixerControlClass *klass)
                               NULL, NULL,
                               g_cclosure_marshal_VOID__UINT,
                               G_TYPE_NONE, 1, G_TYPE_UINT);
+        signals [STREAM_CHANGED] =
+                g_signal_new ("stream-changed",
+                              G_TYPE_FROM_CLASS (klass),
+                              G_SIGNAL_RUN_LAST,
+                              G_STRUCT_OFFSET (GvcMixerControlClass, stream_changed),
+                              NULL, NULL,
+                              g_cclosure_marshal_VOID__UINT,
+                              G_TYPE_NONE, 1, G_TYPE_UINT);
         signals [CARD_ADDED] =
                 g_signal_new ("card-added",
                               G_TYPE_FROM_CLASS (klass),
diff --git a/gvc-mixer-control.h b/gvc-mixer-control.h
index 3f0ea2e..4ba1d3b 100644
--- a/gvc-mixer-control.h
+++ b/gvc-mixer-control.h
@@ -59,6 +59,8 @@ typedef struct
                                         GvcMixerControlState  new_state);
         void (*stream_added)           (GvcMixerControl *control,
                                         guint            id);
+        void (*stream_changed)         (GvcMixerControl *control,
+                                        guint            id);
         void (*stream_removed)         (GvcMixerControl *control,
                                         guint            id);
         void (*card_added)             (GvcMixerControl *control,
-- 
2.1.4

